Woo Installer: Make back button work landing page agnostic

allilevine opened the pull request:
#### Changes proposed in this Pull Request

* Store correct back URL in `back_to` query param

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* On `woocommerce-installation/< site id >` click "Start a new store," then click back. Verify that you return to the same landing page.
* On `/start`, follow the signup wizard and choose "Sell" to enter the seller experience. On "Set up your store" choose "Upgrade," then click back. Verify that you return to the seller experience.
* On `/start`, follow the signup wizard and choose "Sell" to enter the seller experience. On "Set up your store" choose "Upgrade," then click fill the form and hit continue , then click back two times trying to get to the seller experience. Verify that you return to the seller experience.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/60569

allilevine committed:
Store back URL in query param

obenland commented on client/my-sites/woocommerce/landing-page.tsx:
```diff
@@ -44,7 +45,11 @@ const LandingPage: React.FunctionComponent< Props > = ( { siteId } ) => {
 
-		return page( `/start/woocommerce-install/?site=${ wpcomDomain }` );
+		return page(
+			addQueryArgs( `/start/woocommerce-install/?site=${ wpcomDomain }`, {
```
How do you feel about moving the `site` query arg to the object as well?

Rolilink replied:
sounds like a good change 👍🏼

obenland commented on client/my-sites/woocommerce/landing-page.tsx:
```diff
@@ -44,7 +45,11 @@ const LandingPage: React.FunctionComponent< Props > = ( { siteId } ) => {
-		return page( `/start/woocommerce-install/?site=${ wpcomDomain }` );
+		return page(
+			addQueryArgs( `/start/woocommerce-install/?site=${ wpcomDomain }`, {
+				back_to: `/woocommerce-installation/${ wpcomDomain }`,
```
Is there a preference for snake case or camelCase query args in Calypso?

lsl replied:
Not aware of any [calypso standard](https://github.com/Automattic/wp-calypso/blob/1276e7ecb60e62b3784098c411718bafae2e6b7c/docs/coding-guidelines/javascript.md) but WP uses snake_case for query params.

lsl replied:
Just noticed the mix of casing in `steps-pure.js`. I'm guessing this is down to a lot of it never seeing the url so camelCasing made sense, with more stuff hitting the url I think it makes sense to snake case things.

allilevine replied:
I was going with the `redirect_to` case.

Rolilink commented:
Working as expected, something I found while testing a little more is that when we are coming from the `signup` flow and we enter the `woocommerce-install` flow if we continue with the steps like getting into `business-info` and try to get back to the `signup` flow we actually get back to the landing page, is this an issue or is expected?

obenland commented:
That shouldn't happen. Looks like we'll have to carry that back path through the flow.

Rolilink committed:
moves site arg into the arg object

Rolilink committed:
save back_to on context

Rolilink committed:
makes back_to required

lsl commented on client/signup/config/steps-pure.js:
```diff
@@ -735,7 +735,7 @@ export function generateSteps( {
 		'store-address': {
 			stepName: 'store-address',
-			dependencies: [ 'site' ],
+			dependencies: [ 'site', 'back_to' ],
```
```suggestion
			dependencies: [ 'site', 'back_to' ],
			optionalDependencies: [ 'back_to' ],
```

I think we should set the woo landing page as the default for this.

allilevine replied:
Thanks! Fixed here: https://github.com/Automattic/wp-calypso/pull/60892/commits/22515490ed20baec47b50a81cbcd289edd38f0f0

lsl commented on client/signup/config/flows-pure.js:
```diff
@@ -420,7 +420,7 @@ export function generateFlows( {
 			destination: '/',
 			description: 'Onboarding and installation flow for woocommerce on all plans.',
-			providesDependenciesInQuery: [ 'site' ],
+			providesDependenciesInQuery: [ 'site', 'back_to' ],
```
```suggestion
			providesDependenciesInQuery: [ 'site', 'back_to' ],
			optionalDependenciesInQuery: [ 'back_to' ],
```

(See error when hitting `X` in the checkout step)

allilevine replied:
Thanks! Fixed here: https://github.com/Automattic/wp-calypso/pull/60892/commits/22515490ed20baec47b50a81cbcd289edd38f0f0

lsl commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -54,7 +55,8 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
 	const { get, save, update } = useSiteSettings( siteId );
-	const domain = useSelector( ( state ) => getSiteDomain( state, siteId ) );
+
+	const backUrl = signupDependencies.back_to;
```
```suggestion
	const domain = useSelector( ( state ) => getSiteDomain( state, siteId ) );
	const backUrl = signupDependencies?.back_to || `/woocommerce-installation/${ domain }`;
```

Thoughts? I don't know how important it is really but I think forcing a lot of requirements just to get people into this flow is self limiting (e.g. email marketing campaigns might want to skip a click)

lsl replied:
This backUrl should be limited to relative paths.

e.g. This should not work and just ignore it if it's invalid:

```
http://calypso.localhost:3000/start/woocommerce-install/store-address?site=test803879000.wordpress.com&back_to=https://example.com
``` 

Something like this should be enough:
https://github.com/Automattic/wp-calypso/blob/9cf831eeb055e5a65784011fe1337be388658281/client/my-sites/checkout/composite-checkout/lib/leave-checkout.ts#L68

This also blocks less than ideal urls like 
```
http://calypso.localhost:3000/start/woocommerce-install/store-address?site=test803879000.wordpress.com&back_to=javascript:alert(%27sup%27);
```

allilevine replied:
> This backUrl should be limited to relative paths.

Yep looks like we lost that validation in https://github.com/Automattic/wp-calypso/pull/60892/commits/4b2ce06ca91fc575b74bf9e060ecbe2414720aee, let's add that back.

Rolilink replied:
Sorry got a little confused on what that code was doing and removed it my bad there. I think maybe this a good place adding a comment to put some light on what the regexp does.

allilevine replied:
That's a good idea 👍

lsl replied:
Yea, that code really needs to be made a general lib function, there are a few similar/exact copies of it around. 

There is a function that sort of does what we need but it either doesn't work in the right way or is over complicated and hard to figure out if it does what you need it to do (irrc it complicates things by allowing jetpack urls or something like that) - so it doesn't get used. :shrug:

allilevine replied:
Ahhh it would be nice to replace that. 

I added the validation and fallback here: https://github.com/Automattic/wp-calypso/pull/60892/commits/fc3f13356ceaf2b573f3bfeaf83b65e9bb8b70e4

lsl commented:
- [x] On woocommerce-installation/< site id > click "Start a new store," then click back. Verify that you return to the same landing page.
- [ ] On /start, follow the signup wizard and choose "Sell" to enter the seller experience. On "Set up your store" choose "Upgrade," then click back. Verify that you return to the seller experience.
- [ ] On /start, follow the signup wizard and choose "Sell" to enter the seller experience. On "Set up your store" choose "Upgrade," then click fill the form and hit continue , then click back two times trying to get to the seller experience. Verify that you return to the seller experience.

I couldn't get these final two tests working for me,

https://user-images.githubusercontent.com/811776/152913907-18eb3fdd-3ffd-426c-ad57-f490bd6af874.mp4

Can I suggest we get @Rolilink's https://github.com/Automattic/wp-calypso/pull/60866 merged before this one? That might help fix a couple of things regarding how these url's are handled / passed around.

allilevine committed:
Validate backUrl and fall back to default landing page.

allilevine committed:
Add siteSlug to store-features as query arg

allilevine committed:
Make back_to dependency optional.

Rolilink committed:
updates the back_to dependency
Enables passing a back_to url from flow to flow on the woocommerce-install flow

lsl committed:
site/siteSlug on support card

lsl commented on client/signup/steps/woocommerce-install/components/support-card/index.tsx:
```diff
@@ -30,7 +30,7 @@ export default function SupportCard( { backUrl }: { backUrl?: string } ): ReactE
 					<SupportLinkStyle
 						href={ addQueryArgs( '/help/contact', {
-							redirect_to: backUrl || `${ window.location.pathname }?site=${ domain }`,
+							redirect_to: backUrl || `${ window.location.pathname }?siteSlug=${ domain }`,
```
Missed this in #60866 back button on support screen.

lsl reviewed (approved):
I won't claim to know how this is working but it seems to work, LGTM!

- [x] On /start, follow the signup wizard and choose "Sell" to enter the seller experience. On "Set up your store" choose "Upgrade," then click back. Verify that you return to the seller experience.
- [x] On /start, follow the signup wizard and choose "Sell" to enter the seller experience. On "Set up your store" choose "Upgrade," then click fill the form and hit continue , then click back two times trying to get to the seller experience. Verify that you return to the seller experience.
